### PR TITLE
Add autopopulate module via NUSMods share link

### LINKED ISSUE #28: [FEATREQ] Auto populate the data from nusmods timetable
**Is your feature request related to a problem? Please describe.**
As a user, I want to import timetable from nusmods

**Describe the solution you'd like**
- Copy the generated link from nusmods timetable
- nus-timetable optimizer can auto populate the modules (without me manually entering it in)

Do you think this is a good idea?

### PR DESCRIPTION
**Pull request by benclmnt**
Closes #28

### CODE REVIEW

**Comment by frizensami** [comment]
Oh, nice, thanks! Is the PR ready to review?

**Comment by frizensami** [comment]
The concept looks good, and thanks for the help with useEffect and the other housekeeping things. 

A couple of high level points:

1. I'm debating whether the share link population of modules should (a) add any modules that don't currently exist to the list (current behavior), or (b) we should delete the current list of modules and re-populate them with the link. I'm leaning towards (b)'s behavior being more expected for the user. 

2. For UI consistency, we can potentially do two things:
  * Can move the NUSMods share link area to within the Modules tab, either above or below the current manual selection boxes
  * Instead of an `alert`, we can reuse the same disappearing error message system to warn that the link format is invalid. Also I think the code continues even if the share link is invalid at the moment.

Let me know what you think!

**Comment by frizensami** [comment]
Also, could you "allow edits for maintainers" to the PR? I could do some of the small changes.

**Comment by benclmnt** [comment]
Re point 1, I think it is better to keep the current behaviour with an additional clear all modules button. So user can achieve b) by clearing all, then adding all modules. If we were to implement b) directly, then we need to ask for confirmation that it will overwrite their current modules. What do you think?

I agree with point 2. Let me push a fix.

I have allow edits for maintainers 😄

**Comment by benclmnt** [comment]
Oh, and I think another reason in favour of a) is to preserve the "make optional" and the `lessonConstraints`

**Comment by frizensami** [comment]
Okay, the clear all button + current behavior otherwise sounds like a good idea!

File: src/components/ModuleConstraints.tsx
Comment (frizensami) [inline src/components/ModuleConstraints.tsx:187]: I think there's a small bug here with the two error messages (that I didn't account for when writing this style haha). To repro:

* Type "cs2222" (invalid mod) into the modules box but don't submit yet 
* Paste "https://nusmods.com/sem-1/sh" (invalid URL) into the populate module link box but don't submit yet
* Then submit both one after the other
* We expect both error boxes to disappear, since we share the same timeout for both, only 1 box is cleared in the end and the other error stays around forever

**Comment by frizensami** [comment]
Added a couple of small UI changes to make the choice between the link and adding manually more clear, let me know if that seems OK: 

image

File: src/components/ModuleConstraints.tsx
Comment (benclmnt) [reply src/components/ModuleConstraints.tsx]: I had a hacky fix [=, from the user perspective should be good enough if we assume they donot spam the add module / share link button with the wrong module / link HAHAHA

Comment (frizensami) [reply src/components/ModuleConstraints.tsx]: Haha okay let me take a look!

Comment (frizensami) [reply src/components/ModuleConstraints.tsx]: Ok even with decent spamming it works haha :)

Comment (benclmnt) [reply src/components/ModuleConstraints.tsx]: cooll then, I have also added the remove all modules button for completeness

Comment (frizensami) [reply src/components/ModuleConstraints.tsx]: Great! I saw it. Unfortunately the header for the table is removed when a mobile layout is used, so the remove all button also gets removed there. I'll fix it!

**Comment by frizensami** [comment]
I'll just fix the remove-all button for mobile layouts and then this should be good to go

**Comment by frizensami** [comment]
Ah i spotted a bug where if the link contains an invalid module, e.g.

https://nusmods.com/timetable/sem-1/share?CS101S=REC:08A,TUT:09D,LEC:1&CS3230=LEC:1,TUT:12
(CS101S instead of CS1101S)

the optimizer gets into an infinite loop of making API calls. I'll also fix this one, probably a small issue.

**Comment by frizensami** [comment]
OK I fixed everything else, but the bug has something to do with the useEffect-triggered function to check if the new list of mods exist. Could you take a look? 

Something to do with an infinite loop after the call for one of the mods returns false

```
    useEffect(() => {
        for (const module of modules) {
            if (!module.json) {
                // populate json field and then trigger a re-render
                populateModuleConstraintJsonField(module).then(_ => onModulesChange([...modules]));
            }
        }
    }, [modules]);

```

**Comment by benclmnt** [comment]
Pushed a fix, hopefully it works now

**Comment by benclmnt** [comment]
Pushed a fix + reduce number of re-renders :)

**Review by frizensami — APPROVED** [review]
LGTM! Thanks for the help @benclmnt!

**Comment by frizensami** [comment]
I'll deploy this in the next couple of hours :)

**Comment by frizensami** [comment]
Deployed!

**Pull request merged**
